### ISSUE TITLE
TOB-K8S-021: Improper fetching of PIDs allows incorrect cgroup movement

### ISSUE THREAD

**Issue by cji** [kubernetes/kubernetes#81124]
This issue was reported in the [Kubernetes Security Audit Report](https://github.com/kubernetes/community/blob/master/wg-security-audit/findings/Kubernetes%20Final%20Report.pdf)

**Description**
Kubelet fetches a process’s PID by checking the process name when the pidfile file doesn’t exist. This fallback mechanism doesn’t validate the target process allowing an attacker to spoof it.

Kubelet’s container manager first tries to use a hardcoded pidfile path to retrieve the PIDs of important processes, like dockerd and docker-containerd (Figure 1).

```
const (
	// (...)
	dockerProcessName     = "docker"
	dockerPidFile         = "/var/run/docker.pid"
	containerdProcessName = "docker-containerd"
	containerdPidFile     = "/run/docker/libcontainerd/docker-containerd.pid"
)
```
Figure 16.1: Hardcoded pidfiles paths.

If these hard-coded files do not exist, then the container manager will fall back to scanning `/proc/<pid>/cmdline`. If the pidfile is not there, and this happens to be the case on our setup, an attacker may move an unprivileged, cgroups-limited process on the host to manager’s cgroup and set its OOM-killer badness heuristic to -999, making the process less likely to be killed in out-of-memory scenarios.

**Exploit Scenario**
Alice has kubelet running on a machine where the docker-containerd pidfile is located in a different path than the hardcoded one. Eve gains access to Alice’s machine and notices kubelet is running. Eve’s current cgroup limits her too much for what she is attempting to perform on the machine, so she uses kubelet’s cgroup monitoring to migrate her process into an elevated cgroup.

Eve can execute the attack by simply renaming her malicious process to docker-containerd, launching it and waiting five minutes (so kubelet moves it to the cgroup and sets the OOM-killer badness heuristic).

**Recommendation**
Use only pidfiles and let users configure paths to them. This relates to issue #34066. If enforcing the use of pidfiles only is not an option, consider adding checks to identify whether given PID is really the targeted process and warn users that the pidfile hasn’t been found.

**Anything else we need to know?**:

See #81146 for current status of all issues created from these findings.

The vendor gave this issue an ID of TOB-K8S-021 and it was finding 16 of the report.

The vendor considers this issue Medium Severity.

To view the original finding, begin on page 50 of the [Kubernetes Security Review Report](https://github.com/kubernetes/community/blob/master/wg-security-audit/findings/Kubernetes%20Final%20Report.pdf)

**Environment**:

- Kubernetes version: 1.13.4

**Comment by tallclair** [comment]
Note that we already have a safety measure of requiring the process to be in the HostPID namespace: https://github.com/kubernetes/kubernetes/blob/f3a03f71af11e7ac10e74db5b94c9597d7cbe946/pkg/kubelet/cm/container_manager_linux.go#L750-L757

**Comment by disconnect3d** [comment]
Hey, as I initially found this issue I am happy to give some more details here :).

> Note that we already have a safety measure of requiring the process to be in the HostPID namespace:

Yup, the problem relates only to processes on the host - while it is there in the description, it could be a bit more explicit (e.g. why exactly is it "on the host").

**Also while moving a process to a manager's cgroup (usually or always systemd's) on the host is probably not that scary, if this issue is combined with https://github.com/kubernetes/kubernetes/issues/81113, one gets a privileged\* container escape.**

This is because the manager's cgroup, to which the process is moved - `/systemd/system.slice` - has full access to any devices on the machine. As a result, triggering the attack gives a privileged\* container full (RW) disk device access, even though it can't mount the disk device. This has been shown in the Appendix D of the report.

The full attack scenario of this issue and #81113 would look like this:
1. Spawn an unprivileged and fake `docker-containerd` process on the host.
2. Find out the exact time when `kubelet` moves important processes (`docker` and `docker-containerd` iirc) to manager's cgroup and sets their OOM score. This is needed later to hit the race condition. 
3. Prepare a pid-reuse attack such that after `docker-containerd` dies, a new process will get its PID.
4. Kill the fake `docker-containerd` process and spawn a privileged (uid=0) process in the container to trigger the PID-reuse attack and race condition/TOCTOU from https://github.com/kubernetes/kubernetes/issues/81113. **This needs to happen just before the cgroup move and since the whole kubelet action (when it moves important processes to manager's cgroup and sets their OOM score) happens every 5 minutes, this is rather a hard race to win.**
5. Profit(???) - the process inside the container can create the disk device with `mknod` and while it can't `mount` it (because AppArmor disallows it) it can just do raw read/writes to it allowing it to gain full control over the host (e.g. by emulating a mount operation in the user-space via reads/writes to the disk).

\* - by a privileged container, I mean a process inside the container spawned by user with uid=0, not mapped by user namespaces (which are not enabled by default by docker).


---

By the way, the cgroup move can be easily seen with this PoC (run on the host, where `kubelet` is running):
```
vagrant@k8s-1:~$ cp /bin/bash docker-containerd

vagrant@k8s-1:~$ ./docker-containerd

vagrant@k8s-1:~$ cat /proc/self/cgroup
12:perf_event:/
11:cpuset:/
10:freezer:/
9:blkio:/user.slice
8:rdma:/
7:memory:/user.slice
6:cpu,cpuacct:/user.slice
5:pids:/user.slice/user-1000.slice/session-4.scope
4:hugetlb:/
3:devices:/user.slice
2:net_cls,net_prio:/
1:name=systemd:/user.slice/user-1000.slice/session-4.scope
0::/user.slice/user-1000.slice/session-4.scope

vagrant@k8s-1:~$ sleep 5m

vagrant@k8s-1:~$ cat /proc/self/cgroup
12:perf_event:/systemd/system.slice
11:cpuset:/systemd/system.slice
10:freezer:/systemd/system.slice
9:blkio:/systemd/system.slice
8:rdma:/
7:memory:/systemd/system.slice
6:cpu,cpuacct:/systemd/system.slice
5:pids:/systemd/system.slice
4:hugetlb:/systemd/system.slice
3:devices:/systemd/system.slice
2:net_cls,net_prio:/systemd/system.slice
1:name=systemd:/systemd/system.slice
0::/user.slice/user-1000.slice/session-4.scope
```

Ofc this is all available because of a missing pidfile:
```
vagrant@k8s-1:~$ sudo ls -lah /run/docker/libcontainerd/docker-containerd.pid
ls: cannot access '/run/docker/libcontainerd/docker-containerd.pid': No such file or directory
```

Which currently lives in:
```
vagrant@k8s-1:~$ sudo ls -lah /run/docker/containerd/docker-containerd.pid
-rw-r----- 1 root root 4 Aug  9 14:51 /run/docker/containerd/docker-containerd.pid
```

The pidfile probably got moved during one of Docker updates.

PS: This was run on [the kubespray environment we made during the audit (spawned by `vagrant up`)](https://github.com/trailofbits/audit-kubernetes/tree/master/src/environment/kubespray) - and so runs old Kubernetes:
```
vagrant@k8s-1:~$ kubelet --version
Kubernetes v1.13.4
```
